v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/compiler/solcjs-wrapper.ts

```diff
@@ -0,0 +1,138 @@
+// This wrapper was created by extracting the parts of the solc-js package
+// (https://github.com/ethereum/solc-js) that we need to perform compilation.
+
+import * as semver from "semver";
+
+interface Solc {
+  cwrap<T>(ident: string, returnType: string | null, argTypes: string[]): T;
+
+  // eslint-disable-next-line @typescript-eslint/naming-convention -- this is a C function
+  _solidity_reset?: Reset | null;
+  // eslint-disable-next-line @typescript-eslint/naming-convention -- this is a C function
+  _solidity_version?: Version | null;
+  // eslint-disable-next-line @typescript-eslint/naming-convention -- this is a C function
+  _version?: Version | null;
+  // eslint-disable-next-line @typescript-eslint/naming-convention -- this is a C function
+  _compileStandard?: Compile | null;
+  // eslint-disable-next-line @typescript-eslint/naming-convention -- this is a C function
+  _solidity_compile?: Compile | null;
+}
+
+type Reset = () => string;
+type Version = () => string;
+type Compile = (
+  input: string,
+  callbackPtr: number | null,
+  callbackContextPtr?: null,
+) => string;
+
+export interface SolcWrapper {
+  compile: CompileWrapper;
+}
+
+export type CompileWrapper = (input: string) => string;
+
+export default function wrapper(solc: Solc): SolcWrapper {
+  const version = bindVersion(solc);
+  const isVersion6OrNewer = semver.gte(versionToSemver(version()), "0.6.0");
+  const reset = bindReset(solc);
+  const compile = bindCompile(solc, isVersion6OrNewer);
+
+  if (compile === undefined) {
+    // eslint-disable-next-line no-restricted-syntax -- should we use a HardhatError here?
+    throw new Error(
+      'Could not find the "compile" function in the solc library',
+    );
```

Comment:
I'm wondering if we should throw a `HardhatError` here?

Reply:
I'd say yes, unless we expect this to never happen. In that case, I'd use an invariant error.